src/Covenant/Test.hs

```diff
@@ -93,3 +163,574 @@ instance Arbitrary Concrete where
+  labelOptic = lens (\(DataGen a _ _ _ _) -> a) (\(DataGen _ b c d e) a -> DataGen a b c d e)
+
+instance
+  (k ~ A_Lens, a ~ Set ConstructorName, b ~ Set ConstructorName) =>
+  LabelOptic "constructors" k DataGen DataGen a b
+  where
+  {-# INLINEABLE labelOptic #-}
+  labelOptic = lens (\(DataGen _ b _ _ _) -> b) (\(DataGen a _ c d e) b -> DataGen a b c d e)
+
+instance
+  (k ~ A_Lens, a ~ ScopeBoundary, b ~ ScopeBoundary) =>
+  LabelOptic "currentScope" k DataGen DataGen a b
+  where
+  {-# INLINEABLE labelOptic #-}
+  labelOptic = lens (\(DataGen _ _ c _ _) -> c) (\(DataGen a b _ d e) c -> DataGen a b c d e)
+
+instance
+  (k ~ A_Lens, a ~ Map ScopeBoundary Word32, b ~ Map ScopeBoundary Word32) =>
+  LabelOptic "boundVars" k DataGen DataGen a b
+  where
+  {-# INLINEABLE labelOptic #-}
+  labelOptic = lens (\(DataGen _ _ _ d _) -> d) (\(DataGen a b c _ e) d -> DataGen a b c d e)
+
+instance
+  (k ~ A_Lens, a ~ Map TyName (Count "tyvar"), b ~ Map TyName (Count "tyvar")) =>
+  LabelOptic "arities" k DataGen DataGen a b
+  where
+  {-# INLINEABLE labelOptic #-}
+  labelOptic = lens (\(DataGen _ _ _ _ e) -> e) (\(DataGen a b c d _) e -> DataGen a b c d e)
+
+{-  Monadic stack for generating monomorphic datatype declarations. Not every generator uses every part of the state, but
+    it ought to suffice for generating *any* datatype declaration we choose.
+
+    In theory this could be a reader but it becomes super awkward to work, StateT is easier
+
+    While we don't have any generators for polymorphic `ValT`s yet, the scope stuff will be necessary there.
+-}
+newtype DataGenM a = DataGenM (GenT (State DataGen) a)
+  deriving newtype (Functor, Applicative, Monad)
+  deriving (MonadGen) via GenT (State DataGen)
+
+instance MonadState DataGen DataGenM where
+  get = DataGenM $ lift get
+  put = DataGenM . lift . put
+
+{- N.B. We don't need this *yet* but we will need it to generate constructors which take polymorphic functions as arguments.
+-}
+bindVars :: Count "tyvar" -> DataGenM ()
+bindVars count'
+  | count == 0 = crossBoundary
+  | otherwise = do
+      crossBoundary
+      here <- gets (view #currentScope)
+      modify $ over #boundVars (M.insert here $ fromIntegral count)
+  where
+    count :: Int
+    count = review intCount count'
+
+    crossBoundary :: DataGenM ()
+    crossBoundary = modify $ over #currentScope (+ 1)
+
+-- performs action in the deeper scope then resets.
+withBoundVars :: forall (a :: Type). Count "tyvar" -> DataGenM a -> DataGenM a
+withBoundVars count act = do
+  oldScope <- gets (view #currentScope)
+  bindVars count
+  res <- act
+  modify $ set #currentScope oldScope
+  pure res
+
+runDataGenM :: forall (a :: Type). DataGenM a -> Gen a
+runDataGenM (DataGenM ma) = (\x -> evalState x (DataGen M.empty Set.empty 0 M.empty M.empty)) <$> GT.runGenT ma
+
+chooseIntT :: forall (m :: Type -> Type). (MonadGen m) => (Int, Int) -> m Int
+chooseIntT bounds = GT.liftGen $ chooseInt bounds
```

Comment:
This feels useful enough to expose publically from this module I think. Name it `chooseInt` though, since it's the same functionality, just more general.